### PR TITLE
MRG: Add BEM surface extraction

### PR DESCRIPTION
**Pull request by hoechenberger**
Add BEM surface generation via watershed algorithm.

### CODE REVIEW

File: config.py [hunk 1]
```diff
@@ -744,10 +744,18 @@
 """
 
 ###############################################################################
-# SOURCE SPACE PARAMETERS
-# -----------------------
+# SOURCE ESTIMATION PARAMETERS
+# ----------------------------
 #
 
+recreate_bem_surfaces: bool = False
```
Comment (agramfort) [inline config.py:751]: we have 2 ways to compute BEM surfaces: watershed and flash
that should be one parameter.

then we need another parameter that says if we force recompute.

wdyt?

Comment (hoechenberger) [reply config.py:751]: yes absolutely, will add

Comment (hoechenberger) [reply config.py]: I've included FLASH MRI support now

File: scripts/source/01-make_bem_surfaces.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            msg = 'Skipping surface extraction as requested in configuration.'
+            logger.info(gen_log_message(step=10, message=msg))
+            return
+
+    if bem_from_flash:
+        msg = 'Creating BEM surfaces from FLASH MRI images'
+        logger.info(gen_log_message(step=10, message=msg))
+        mne.bem.make_flash_bem(subject=fs_subject,
+                               subjects_dir=fs_subjects_dir,
+                               copy=True,  # XXX Revise!
```
Comment (agramfort) [inline scripts/source/01-make_bem_surfaces.py:58]: why revise?

File: scripts/source/01-make_bem_surfaces.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+@failsafe_run(on_error=on_error)
+def make_bem(subject):
+    fs_subject = config.get_fs_subject(subject)
+    fs_subjects_dir = config.get_fs_subjects_dir()
+    mri_dir = Path(fs_subjects_dir) / fs_subject / 'mri'
+    bem_dir = Path(fs_subjects_dir) / fs_subject / 'bem'
+    watershed_bem_dir = bem_dir / 'watershed'
+    flash_bem_dir = bem_dir / 'flash'
+    flash_dir = mri_dir / 'flash' / 'parameter_maps'
```
Comment (agramfort) [inline scripts/source/01-make_bem_surfaces.py:30]: hum technically we should write the flash images in the BIDS dataset in the anat folder and use these. As written it can only work for the sample dataset.

File: scripts/source/01-make_bem_surfaces.py (hunk 2 above)
Comment (hoechenberger) [reply scripts/source/01-make_bem_surfaces.py:58]: Because I'm not sure if we want `copy=True`… What's your opinion?

File: .circleci/config.yml [hunk 4]
```diff
@@ -76,6 +87,7 @@ jobs:
                which python
                git-annex version
                datalad --version
+               freesurver --version
```
Comment (agramfort) [inline .circleci/config.yml:90]:
```suggestion
               freesurfer --version
```

File: config.py [hunk 5]
```diff
@@ -788,13 +817,6 @@
     ```
 """
 
-smooth: Optional[int] = 10
-"""
-Number of iterations for the smoothing of the surface data.
-If None, smooth is automatically defined to fill the surface
-with non-zero values. The default is spacing=None.
-"""
```
Comment (agramfort) [inline config.py:796]: why did you remove this? it's still useful for group morphing I think.

File: scripts/source/01-make_bem_surfaces.py (hunk 2 above)
Comment (agramfort) [reply scripts/source/01-make_bem_surfaces.py:58]: you should stick to True as symlinks are problematic on windows and when you upload to openneuro for example

File: scripts/source/01-make_bem_surfaces.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                               copy=True,  # XXX Revise!
+                               overwrite=True,
+                               show=show)
+    else:
+        msg = ('Creating BEM surfaces from T1-weighted MRI images using '
+               'watershed algorithm')
+        logger.info(gen_log_message(step=10, message=msg))
+        mne.bem.make_watershed_bem(subject=fs_subject,
+                                   subjects_dir=fs_subjects_dir,
+                                   copy=True,  # XXX Revise!
```
Comment (agramfort) [inline scripts/source/01-make_bem_surfaces.py:67]: idem. Keep it true

**Review by agramfort — COMMENTED** [review]
does it work for you as is?

File: scripts/source/01-make_bem_surfaces.py (hunk 6 above)
Comment (hoechenberger) [reply scripts/source/01-make_bem_surfaces.py:67]: Good point, thanks

File: config.py (hunk 5 above)
Comment (hoechenberger) [reply config.py:796]: Yes, but in fact, we didn't actually make use of this param. I'd rather add it back at some later time with proper support & tests

**Comment by hoechenberger** [comment]
> does it work for you as is?

Still testing, my current setup is a bit messy ;) so need to clean up first have something reproducible!

**Comment by hoechenberger** [comment]
Doc build will fail until https://github.com/conda-forge/mkdocs-material-feedstock/pull/44 has been merged.

**Comment by hoechenberger** [comment]
@agramfort Tests are passing, BEM is created from FLASH images for ds000248.
https://1478-167957354-gh.circle-artifacts.com/0/reports/ds000248/sub-01_task-audiovisual_report.html#17

**Review by agramfort — APPROVED** [review]
nice !

**Comment by agramfort** [comment]
looks great @hoechenberger !

I see that our reports become pretty huge. For example https://1478-167957354-gh.circle-artifacts.com/0/reports/ds000248/sub-01_task-audiovisual_report.html#16 is 55MB ...

I am wondering if we could use .svg.gz rather than PNGs in the html for certain plots (line plots etc.)

if the size is mainly due to the BEM slices maybe we can use jpg rather than PNG?

food for thoughts @hoechenberger

**Comment by hoechenberger** [comment]
> I am wondering if we could use .svg.gz rather than PNGs in the html for certain plots (line plots etc.)
> 
> if the size is mainly due to the BEM slices maybe we can use jpg rather than PNG?

I think it's mainly the BEMs, but I have some ideas on how to revamp reports. Should be the next thing on my todo list anyway, I suppose :)

**Comment by agramfort** [comment]
great !

feel free to merge if happy!

>

**Comment by jasmainak** [comment]
If you want to revamp the report and optimize the memory for BEM computation, this might be a good place to start:

https://github.com/mne-tools/mne-python/issues/5425
